### ISSUE TITLE
$.cookie is not a function

### ISSUE THREAD

**Issue by johnantoni** [okta/okta-signin-widget#380]
Hi, getting **$.cookie is not a function** in the @okta/okta-signin-widget 2.5.0

![screen shot 2018-01-22 at 1 53 12 pm](https://user-images.githubusercontent.com/15571/35238509-a2720acc-ff7b-11e7-94a6-939ec778c952.png)

Here's my source:

```
import React, { Component } from 'react';
import OktaSignIn from '@okta/okta-signin-widget';
import '@okta/okta-signin-widget/dist/css/okta-sign-in.min.css';
import '@okta/okta-signin-widget/dist/css/okta-theme.css';

import config from './../.config';

export default class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.signIn = new OktaSignIn({
      /**
       * Note: when using the Sign-In Widget for an OIDC flow, it still
       * needs to be configured with the base URL for your Okta Org. Here
       * we derive it from the given issuer for convenience.
       */
      baseUrl: config.oidc.issuer.split('/oauth2')[0],
      clientId: config.oidc.clientId,
      redirectUri: config.oidc.redirectUri,
      // logo: '/react.svg',
      i18n: {
        en: {
          'primaryauth.title': 'Sign in to React & Company',
        },
      },
      authParams: {
        responseType: ['id_token', 'token'],
        issuer: config.oidc.issuer,
        display: 'page',
        scopes: config.oidc.scope.split(' '),
      },
    });
  }
  componentDidMount() {
    this.signIn.renderEl(
      { el: '#sign-in-widget' },
      () => {
        /**
         * In this flow, the success handler will not be called beacuse we redirect
         * to the Okta org for the authentication workflow.
         */
      },
      (err) => {
        throw err;
      },
    );
  }
  render() {
    return (
      <div>
        <div id="sign-in-widget" />
      </div>
    );
  }
}

```

Any ideas?

looks related to a similar error in december which was clossed https://github.com/okta/okta-signin-widget/issues/215

**Comment by johnantoni** [comment]
Here's my source:

```
import React, { Component } from 'react';
import OktaSignIn from '@okta/okta-signin-widget';
import '@okta/okta-signin-widget/dist/css/okta-sign-in.min.css';
import '@okta/okta-signin-widget/dist/css/okta-theme.css';

import config from './../.config';

export default class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.signIn = new OktaSignIn({
      /**
       * Note: when using the Sign-In Widget for an OIDC flow, it still
       * needs to be configured with the base URL for your Okta Org. Here
       * we derive it from the given issuer for convenience.
       */
      baseUrl: config.oidc.issuer.split('/oauth2')[0],
      clientId: config.oidc.clientId,
      redirectUri: config.oidc.redirectUri,
      // logo: '/react.svg',
      i18n: {
        en: {
          'primaryauth.title': 'Sign in to React & Company',
        },
      },
      authParams: {
        responseType: ['id_token', 'token'],
        issuer: config.oidc.issuer,
        display: 'page',
        scopes: config.oidc.scope.split(' '),
      },
    });
  }
  componentDidMount() {
    this.signIn.renderEl(
      { el: '#sign-in-widget' },
      () => {
        /**
         * In this flow, the success handler will not be called beacuse we redirect
         * to the Okta org for the authentication workflow.
         */
      },
      (err) => {
        throw err;
      },
    );
  }
  render() {
    return (
      <div>
        <div id="sign-in-widget" />
      </div>
    );
  }
}

```

**Comment by johnantoni** [comment]
looks like the issue is still around with 2.5.0

managed to resolve with using @CharlieKuharski's fix below

`const OktaSignIn = require('@okta/okta-signin-widget/dist/js/okta-sign-in.min.js');`

I think part of the problem is the jquery.cookie lib in package.json which is depreciated and needs to be replaced with the js-cookie library https://github.com/js-cookie/js-cookie

**Comment by StanMarian** [comment]
Hello,
Seems like the issue is still around in 2.6.0 and  @CharlieKuharski's fix doesn't work for me anymore. It worked 4 days ago (20.01.2018). 
If I try `const OktaSignIn = require('@okta/okta-signin-widget/dist/js/okta-sign-in.min.js');` I get 

> Error: Zone.js has detected that ZoneAwarePromise `(window|global).Promise` has been overwritten.

Is there another solution to this? Maybe a quick 2.6.1 just for replacing `jquery.cookie` with `js-query` ?
Thank you.

**Comment by johnantoni** [comment]
agree @StanMarian, opened a pull request just not having much luck getting the tests to pass

If you've got a moment to look over and see where I've screwed up that'd be excellent https://github.com/okta/okta-signin-widget/pull/382

Thanks,

**Comment by StanMarian** [comment]
Firstly, I've managed to find a workaround using @CharlieKuharksi's fix:
- in `polyfills.ts` one has to add the import like this:
`import '@okta/okta-signin-widget/dist/js/okta-sign-in-no-jquery.js';
import 'zone.js/dist/zone';  // Included with Angular CLI.` (before zone)

@johnantoni I'll try to have a look

**Comment by StanMarian** [comment]
@johnantoni I've looked over the tests you've written and I think the problem comes from the fact that `$.cookie` was used as a getter and as a setter and now it has been replaced (in tests) only by `Cookies.get`. I hope this will help you.

**Comment by mtschammer** [comment]
Any update on this? It's preventing our organization from using Okta.

**Comment by johnantoni** [comment]
I put some work into the switch to js-cookie but couldn't get the tests to pass.

If someone else can pick it up that'd be awesome, would myself but got a massive demo to get ready for tomorrow.

https://github.com/okta/okta-signin-widget/pull/382

**Comment by agustincelentano** [comment]
We are about to leave okta because of this bug, is there any update? :(

I am using Angular with Sign In Widget

**Comment by robertjd** [comment]
Hi everyone, sorry that this is still causing a problem.  Can you try taking a look at our new sample application, specifically the Login component found in the "custom login" example:

https://github.com/okta/samples-js-react/blob/master/custom-login/src/Login.jsx

It looks pretty much identical to the code that's pasted above, but the difference is that we do the wildcard import style:

```
import * as OktaSignIn from '@okta/okta-signin-widget';
```

Can you try this to see if it fixes the problem?

**Comment by johnantoni** [comment]
Hi @robertjd , tried that and getting "OktaSignIn is not a constructor"

<img width="545" alt="screen shot 2018-02-28 at 3 41 36 pm" src="https://user-images.githubusercontent.com/15571/36812077-eebbc2bc-1c9d-11e8-9f3d-b1e2d571bba6.png">

`import React, { Component } from "react";
import * as OktaSignIn from '@okta/okta-signin-widget';

// const OktaSignIn = require('@okta/okta-signin-widget/dist/js/okta-sign-in.min.js');

import '@okta/okta-signin-widget/dist/css/okta-sign-in.min.css';
import '@okta/okta-signin-widget/dist/css/okta-theme.css';

import config from './../.config';

class LoginPage extends Component {
  constructor(props) {
    super(props);
    this.signIn = new OktaSignIn({
      /**
       * Note: when using the Sign-In Widget for an OIDC flow, it still
       * needs to be configured with the base URL for your Okta Org. Here
       * we derive it from the given issuer for convenience.
       */
      baseUrl: config.oidc.issuer.split('/oauth2')[0],
      clientId: config.oidc.clientId,
      redirectUri: config.oidc.redirectUri,
      logo: "test",
      i18n: {
        en: {
          'primaryauth.title': 'Sign In',
        },
      },
      authParams: {
        responseType: ['id_token', 'token'],
        issuer: config.oidc.issuer,
        display: 'page',
        scopes: config.oidc.scope.split(' '),
      },
    })
  }

  signInWidget = () => {
    this.signIn.renderEl(
      { el: '#sign-in-widget' },
      () => {
        /**
         * In this flow, the success handler will not be called beacuse we redirect
         * to the Okta org for the authentication workflow.
         */
      },
      (err) => {
        throw err;
      },
    );
  }

  componentDidMount() {
    this.signInWidget()
  }

  render() {
    return (
      <div id="sign-in-widget" />
    )
  }
}

export default LoginPage;
`

**Comment by oklas** [comment]
Using jquery in react app is bad idea. Using `/dist/js/okta-sign-in-no-jquery.js` is very bad idea. It include almost a Megabyte overhead to bundle which will be loaded each time. Include jquery itself from any CDN is better because is a chance that it already in browser cache and downloading may be omited.

Is any progress about to avoid to use jquery in modern applications?

**Comment by oklas** [comment]
This question is opened at least since Apr 2017 #215.

**Comment by oklas** [comment]
Using jquery in react app is dubious idea. Using /dist/js/okta-sign-in-no-jquery.js is very bad idea. It include almost a Megabyte overhead to bundle which will be loaded each time. Include jquery itself from any CDN is better because is a chance that it already in browser cache and downloading may be omited.

Is any progress about to avoid to use jquery in modern applications?

**Comment by oklas** [comment]
Using jquery in react app is dubious idea. Using /dist/js/okta-sign-in-no-jquery.js is very bad idea. It include almost a Megabyte overhead to bundle which will be loaded each time. Include jquery itself from any CDN is better because is a chance that it already in browser cache and downloading may be omited.

Is any chance about to avoid to use jquery in modern applications?

**Comment by pydanny** [comment]
This is a complete blocker for us to use Okta. Any movement on this?

**Comment by s-toussaint** [comment]
Hi,

I have followed the guide in the following page [https://developer.okta.com/code/angular/okta_angular_sign-in_widget](url) and I ran into the same issue: 

zone.js:196 Uncaught TypeError: $.cookie is not a function
    at Object.fn.getCookieUsername (okta-sign-in.entry.js:38958)
    at constructor.props (okta-sign-in.entry.js:38780)
    at Function._.result (underscore.js:1370)
    at constructor (okta-sign-in.entry.js:17849)
    at constructor (okta-sign-in.entry.js:17439)
    at new constructor (okta-sign-in.entry.js:38833)
    at new constructor (okta-sign-in.entry.js:38306)
    at child.render (okta-sign-in.entry.js:34512)
    at _fulfilled (q.js:834)
    at eval (q.js:863)

Let me know how I can to help you solve this issue.

Thank you.

**Comment by s-toussaint** [comment]
While trying to gather more information for my above comment, I did the following change in the import
`import * as OktaSignIn from '@okta/okta-signin-widget/dist/js/okta-sign-in.min.js';`
instead of
`import * as OktaSignIn from '@okta/okta-signin-widget';`
and it works!

However, once I signed in from the /login page, I was redirected to the "normal" okta sign-in page so I changed the following line in LoginComponent from
`this.signIn.loginRedirect({ sessionToken: res.session.token });`
to (**added '/home' uri**)
`this.signIn.loginRedirect('/home', { sessionToken: res.session.token });`

I have also changed the logout call from
`<button *ngIf="signIn.isAuthenticated()" (click)="logout()"> Logout </button>`
to (**added '/login' uri**)
`<button *ngIf="signIn.isAuthenticated()" (click)="logout('/login')"> Logout </button>`
so I get redirected to the login page once logged out.

On a side note I have noticed with the widget is that the company logo is not automatically displayed. In order to have it, you need to add the logo property to the config:
```
widget = new OktaSignIn({
    baseUrl: 'https://dev-xxxxxxxxxxxx.oktapreview.com',
    logo: '/assets/img/logo.png',
  });
```

Hope this helps!

**Comment by rdegges** [comment]
@nbarbettini any chance of getting this one looked at too?

**Comment by jmelberg-okta** [comment]
Hi all,

We're happy to say this fix has now been included with the `2.10.0` version.

Please feel free to reopen is issue if any problems arise.
